### PR TITLE
UIP-3068 Add ability to exclude directories when checking for infractions

### PR DESCRIPTION
**Pull request by jacehensley-wf**
## Ultimate Problem
https://github.com/Workiva/dependency_validator/issues/8

## Solution
- Add the `--exclude-dir` flag

## Testing Suggestions
- Verify tests pass

## Possible Areas of Regression
- Error reporting

---
FYA: @greglittlefield-wf @maxwellpeterson-wf @toddbeckman-wf 
FYI: @evanweible-wf

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Module Findings
No security relevant content detected. Please review for security relevance and request security review as needed.

File: lib/src/utils.dart [hunk 1]
```diff
@@ -29,11 +29,17 @@ final Logger logger = new Logger('dependency_validator');
...
   if (!FileSystemEntity.isDirectorySync(dirPath)) return const [];
-  return new Directory(dirPath)
-      .listSync(recursive: true)
-      .where((entity) => entity is File && !entity.path.contains('/packages/') && entity.path.endsWith('.dart'));
+
+  return new Directory(dirPath).listSync(recursive: true).where((entity) {
+    if (entity is! File) return false;
+    if (entity.path.contains('/packages/')) return false;
+    if (!entity.path.endsWith('.dart')) return false;
+    if (excludedDirs.any(entity.path.startsWith)) return false;
```
Comment (greglittlefield-wf) [inline lib/src/utils.dart]: This could cause false positives if a directory is a substring of the excluded dir and the user forgets a trailing slash.

You can use [`path.isWithin`](https://www.dartdocs.org/documentation/path/1.5.1/path/isWithin.html) to handle this.

File: lib/src/utils.dart [hunk 2]
```diff
@@ -29,11 +29,17 @@ final Logger logger = new Logger('dependency_validator');
...
-Iterable<File> listDartFilesIn(String dirPath) {
+Iterable<File> listDartFilesIn(String dirPath, List<String> excludedDirs) {
   if (!FileSystemEntity.isDirectorySync(dirPath)) return const [];
-  return new Directory(dirPath)
-      .listSync(recursive: true)
-      .where((entity) => entity is File && !entity.path.contains('/packages/') && entity.path.endsWith('.dart'));
+
+  return new Directory(dirPath).listSync(recursive: true).where((entity) {
+    if (entity is! File) return false;
+    if (entity.path.contains('/packages/')) return false;
```
Comment (greglittlefield-wf) [inline lib/src/utils.dart]: Path #nit; it would be more portable to do:
```dart
if (path.split(entity.path).contains('packages')) return false;
```

https://www.dartdocs.org/documentation/path/1.5.1/path/split.html

File: test/executable_test.dart [hunk 3]
```diff
@@ -66,6 +68,13 @@ void main() {
         expect(result.stderr, contains('These packages are used in lib/ but are not dependencies:'));
         expect(result.stderr, contains('yaml'));
       });
+
+      test('expect when the lib directory is excluded', () {
```
Comment (greglittlefield-wf) [inline test/executable_test.dart]: Typo: should be `except`

File: lib/src/utils.dart (hunk 1 above)
Comment (jacehensley-wf) [reply lib/src/utils.dart]: whoa... Dart never ceases to amaze me

File: lib/src/utils.dart [hunk 4]
```diff
@@ -34,9 +35,9 @@ Iterable<File> listDartFilesIn(String dirPath, List<String> excludedDirs) {
 
   return new Directory(dirPath).listSync(recursive: true).where((entity) {
     if (entity is! File) return false;
-    if (entity.path.contains('/packages/')) return false;
-    if (!entity.path.endsWith('.dart')) return false;
-    if (excludedDirs.any(entity.path.startsWith)) return false;
+    if (entity.path.split(p.separator).contains('packages')) return false;
```
Comment (greglittlefield-wf) [inline lib/src/utils.dart]: This is probably fine but in the future I'd recommend using `p.split`, and not `.split(p.separator)`.

For example:
```dart
// Multiple slashes should collapse to one in most cases
print('foo//bar'.split(p.separator)); // [foo, , bar]
print(p.split('foo//bar'));           // [foo, bar]

// In Windows, you can mix slashes
print(r'foo/bar\baz'.split(p.windows.separator)); // [foo/bar, baz]
print(p.windows.split(r'foo/bar\baz'));           // [foo, bar, baz]
```

Paths have so many edge-cases, and it's usually better to not make any assumptions and let the `path` package take care of it for you.

**Comment by maxwellpeterson-wf** [comment]
+10
- Tests pass.
- New tests cover additional CLI options.

**Comment by maxwellpeterson-wf** [comment]
QA +1
- [x] Dev +1's
- [x] Dev/QA +10 with detail of what was tested
  - https://github.com/Workiva/dependency_validator/pull/15#issuecomment-368906834
- [x] Tests created/updated 
- [x] All tests pass

@Workiva/release-management-pp
